File: serverless/fetchProducts.rb
require 'json'
require 'shopify_api'

def fetchProducts(event:, context:)
  shop_url = "https://#{ENV['API_KEY']}:#{ENV['PASSWORD']}@#{ENV['SHOP_NAME']}.myshopify.com"
  ShopifyAPI::Base.site = shop_url
  ShopifyAPI::Base.api_version = '2020-04'
  
  page = 1
  count = ShopifyAPI::Product.count
  puts "Found #{count} products in Shopify"
  if count > 0
    product_details = []
    page += count.divmod(50).first
    while page > 0
      products = ShopifyAPI::Product.find(:all, params: {limit: 50})
      
      product_details += products.map {|p| {
        id: p.id,
        title: p.title,
        price: p.variants[0].price,
        images: p.images[0].src
      }}
      page -= 1
    end
  end
  puts "Processing #{product_details.length} products from Shopify..."
  
  headers = {
    "Access-Control-Allow-Origin": "*"
  }

  { statusCode: 200, headers: headers, body: JSON.generate(product_details) }
end

File: serverless/fetchProducts.js
'use strict';
const Shopify = require('shopify-api-node');
const AWS = require('aws-sdk');
// Set the region and create the DynamoDB service object
AWS.config.update({region: 'ap-northeast-1'});
const ddb = new AWS.DynamoDB({apiVersion: '2012-08-10'});

module.exports.fetchProducts = async (event, context, callback) => {
  const params = {
    TableName: 'Shop',
    Key: {
      shopName: {
        S: event["queryStringParameters"]["shopName"]
      }
    }
  }

  let res = {}

  await ddb.getItem(params, function(err, data) {
    if (err) {
      console.log("err", err);
    } else {
      res = data
    }
  }).promise();

  const shopify = await new Shopify({
    shopName: event["queryStringParameters"]["shopName"],
    accessToken: res.Item.accessToken.S
  });

  const products = await shopify.product.list({ limit: 10 });

  const response = {
    statusCode: 200,
    headers: {
      "Access-Control-Allow-Origin": "*",
      "Access-Control-Allow-Headers": "Origin, X-Requested-With, Content-Type, Accept, Authorization",
      "Access-Control-Allow-Methods": "GET, PUT, POST, DELETE, PATCH, OPTIONS"
    },
    body: JSON.stringify({products}),
  };

  return response;
};
